File: swift/examples/GoodService.swift
import Foundation

// ✅ Domain model: immutable value type with validation
struct Money: Equatable {
    let amount: Decimal
    let currency: String

    init(amount: Decimal, currency: String) throws {
        guard amount >= 0 else {
            throw ValidationError.negativeAmount(amount)
        }
        guard currency.count == 3 else {
            throw ValidationError.invalidCurrency(currency)
        }
        self.amount = amount
        self.currency = currency
    }
}

// ✅ Typed errors — never use NSError or generic Error directly
enum ValidationError: LocalizedError {
    case negativeAmount(Decimal)
    case invalidCurrency(String)

    var errorDescription: String? {
        switch self {
        case .negativeAmount(let amount):
            return "Amount must be non-negative, got: \(amount)"
        case .invalidCurrency(let code):
            return "Currency must be ISO 4217 (3 chars), got: \(code)"
        }
    }
}

enum PaymentError: LocalizedError {
    case orderNotFound(String)
    case orderNotPayable(String, OrderStatus)
    case cardDeclined(String)
    case serviceUnavailable(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .orderNotFound(let id):           return "Order \(id) not found"
        case .orderNotPayable(let id, let s):  return "Order \(id) is not payable (status: \(s))"
        case .cardDeclined(let code):          return "Card declined: \(code)"
        case .serviceUnavailable(let e):       return "Payment service unavailable: \(e)"
        }
    }
}

enum OrderStatus: String {
    case pending, paid, cancelled
}

// ✅ Protocol for dependency injection — enables easy test doubles
protocol OrderRepositoryProtocol {
    func findOrder(byID id: String) async throws -> Order?
    func save(_ order: Order) async throws
}

protocol PaymentGatewayProtocol {
    func charge(amount: Decimal, currency: String, cardToken: String) async throws -> String
}

// ✅ Final class — prevents unintended subclassing
final class OrderPaymentService {
    private let orderRepository: OrderRepositoryProtocol
    private let paymentGateway: PaymentGatewayProtocol

    init(orderRepository: OrderRepositoryProtocol, paymentGateway: PaymentGatewayProtocol) {
        self.orderRepository = orderRepository
        self.paymentGateway = paymentGateway
    }

    /// Process a payment for an existing pending order.
    ///
    /// - Parameters:
    ///   - orderID: The order to charge. Must not be empty.
    ///   - cardToken: Tokenised card from the payment provider.
    /// - Returns: The payment confirmation ID.
    /// - Throws: `PaymentError` describing the failure reason.
    func processPayment(orderID: String, cardToken: String) async throws -> String {
        // ✅ Guard for preconditions — fail fast, keep the happy path flat
        guard !orderID.isEmpty else {
            throw ValidationError.negativeAmount(0) // replace with proper precond error
        }

        guard let order = try await orderRepository.findOrder(byID: orderID) else {
            throw PaymentError.orderNotFound(orderID)
        }

        guard order.status == .pending else {
            throw PaymentError.orderNotPayable(orderID, order.status)
        }

        let paymentID: String
        do {
            paymentID = try await paymentGateway.charge(
                amount: order.total.amount,
                currency: order.total.currency,
                cardToken: cardToken
            )
        } catch let error as PaymentError {
            throw error // re-throw domain errors as-is
        } catch {
            throw PaymentError.serviceUnavailable(underlying: error)
        }

        var updatedOrder = order
        updatedOrder.markAsPaid(paymentID: paymentID)
        try await orderRepository.save(updatedOrder)

        return paymentID
    }
}

// ✅ Value type for domain entity — structs preferred for models in Swift
struct Order {
    let id: String
    let userID: String
    var total: Money
    var status: OrderStatus
    private(set) var paymentID: String?

    mutating func markAsPaid(paymentID: String) {
        self.status = .paid
        self.paymentID = paymentID
    }
}

File: rust/examples/good_service.rs
//! Order payment processing — demonstrates idiomatic Rust patterns.
//!
//! Covered patterns:
//! - `thiserror` for typed, composable errors
//! - Traits for dependency injection (testable)
//! - `async fn` with `tokio`
//! - Newtype for validated value objects
//! - `?` operator for clean error propagation

use std::fmt;
use async_trait::async_trait;
use thiserror::Error;
use uuid::Uuid;

// ✅ Newtype wraps String — prevents mixing order IDs with user IDs at compile time
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct OrderId(Uuid);

impl OrderId {
    pub fn new() -> Self {
        Self(Uuid::new_v4())
    }

    pub fn from_str(s: &str) -> Result<Self, PaymentError> {
        Uuid::parse_str(s)
            .map(Self)
            .map_err(|_| PaymentError::InvalidInput("orderId must be a valid UUID".into()))
    }
}

impl fmt::Display for OrderId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

// ✅ thiserror generates std::error::Error impl — no manual boilerplate
#[derive(Debug, Error)]
pub enum PaymentError {
    #[error("order {0} not found")]
    OrderNotFound(OrderId),

    #[error("order {0} is not payable (status: {1:?})")]
    OrderNotPayable(OrderId, OrderStatus),

    #[error("card declined: {code}")]
    CardDeclined { code: String },

    #[error("payment service unavailable: {0}")]
    ServiceUnavailable(#[source] anyhow::Error),

    #[error("invalid input: {0}")]
    InvalidInput(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OrderStatus {
    Pending,
    Paid,
    Cancelled,
}

#[derive(Debug, Clone)]
pub struct Order {
    pub id: OrderId,
    pub user_id: String,
    pub total_cents: u64,
    pub currency: String,
    pub status: OrderStatus,
}

// ✅ Trait for DI — impl can be swapped for a fake in tests
#[async_trait]
pub trait OrderRepository: Send + Sync {
    async fn find_by_id(&self, id: &OrderId) -> Result<Option<Order>, anyhow::Error>;
    async fn save(&self, order: &Order) -> Result<(), anyhow::Error>;
}

#[async_trait]
pub trait PaymentGateway: Send + Sync {
    async fn charge(
        &self,
        amount_cents: u64,
        currency: &str,
        card_token: &str,
    ) -> Result<String, PaymentError>;
}

// ✅ Service owns trait objects via Arc — shared, thread-safe
pub struct OrderPaymentService {
    orders: std::sync::Arc<dyn OrderRepository>,
    gateway: std::sync::Arc<dyn PaymentGateway>,
}

impl OrderPaymentService {
    pub fn new(
        orders: std::sync::Arc<dyn OrderRepository>,
        gateway: std::sync::Arc<dyn PaymentGateway>,
    ) -> Self {
        Self { orders, gateway }
    }

    /// Process a payment for an existing pending order.
    ///
    /// Returns the payment confirmation ID on success.
    pub async fn process_payment(
        &self,
        order_id: &OrderId,
        card_token: &str,
    ) -> Result<String, PaymentError> {
        if card_token.is_empty() {
            return Err(PaymentError::InvalidInput("card_token must not be empty".into()));
        }

        // ✅ `?` propagates errors without manual match — wraps via From
        let order = self
            .orders
            .find_by_id(order_id)
            .await
            .map_err(|e| PaymentError::ServiceUnavailable(e))?
            .ok_or_else(|| PaymentError::OrderNotFound(order_id.clone()))?;

        if order.status != OrderStatus::Pending {
            return Err(PaymentError::OrderNotPayable(order_id.clone(), order.status.clone()));
        }

        let payment_id = self
            .gateway
            .charge(order.total_cents, &order.currency, card_token)
            .await?;

        // ✅ Only persist after payment succeeds
        let paid_order = Order { status: OrderStatus::Paid, ..order };
        self.orders
            .save(&paid_order)
            .await
            .map_err(|e| PaymentError::ServiceUnavailable(e))?;

        Ok(payment_id)
    }
}
